### ISSUE TITLE
Sending SDP data to a Powercore

### ISSUE THREAD

**Issue by fgoupil-cbc** [bitfocus/companion-module-generic-emberplus#43]
Hi,

Has anyone been able to push an SDP file to a Powercore (or any other Ember+ device I guess) from Companion so that it subscribes to a stream? The Powercore expects carriage returns for each line of the SDP file but Companion only allows single line text and I tried adding \n to encode each line but Companion keeps sending that as clear text from what I can see using Wireshark. I at least know that Companion is able to empty out the receiver by sending an empty line.

Any ideas?

Thanks,

François

**Comment by phillipivan** [comment]
Hi Francois,

What module version are using?

I presume the SDP field is a string parameter, is that correct?

If you download module version 2.4.4 you could create the multiline SDP entry as a variable and try using that in the Set Value String action.

**Comment by phillipivan** [comment]
Great. The built version of the module is attached to the release.

**Comment by phillipivan** [comment]
Have you been able to test this yet?

**Comment by cbjornram** [comment]
Hi Phillip,

I also gave this a go (Companion version 3.5.0-7590 and generic ember module version 2.5.0) and can't get it working either. I also tried the idea with a multi line variable using `\n` for each new line in the SDP, as well as one at the end of the SDP (as required of the Power Core), but with the same result. Just like François, I can clear an SDP by sending an empty string.

Original SDP:

```
v=0
o=- 09420142300001 0 IN IP4 192.168.99.240
s=core01:rav01
t=0 0
a=clock-domain:PTPv2 0
a=ts-refclk:ptp=IEEE1588-2008:00-0B-72-FF-FE-09-88-F6:0
a=mediaclk:direct=0
a=group:DUP ra0 ra1
m=audio 5004 RTP/AVP 98
c=IN IP4 239.99.240.1/16
a=source-filter: incl IN IP4 239.99.240.1 192.168.99.240
a=rtpmap:98 L24/48000/2
a=mid:ra0
a=framecount:48
a=recvonly
a=ptime:1
a=sync-time:0
a=mediaclk:direct=0
m=audio 5004 RTP/AVP 98
c=IN IP4 239.98.240.1/16
a=source-filter: incl IN IP4 239.98.240.1 192.168.98.240
a=rtpmap:98 L24/48000/2
a=mid:ra1
a=framecount:48
a=recvonly
a=ptime:1
a=sync-time:0
a=mediaclk:direct=0

```

Manually edited for custom variable:

```
v=0\no=- 09420142300001 0 IN IP4 192.168.99.240\ns=core01:rav01\nt=0 0\na=clock-domain:PTPv2 0\na=ts-refclk:ptp=IEEE1588-2008:00-0B-72-FF-FE-09-88-F6:0\na=mediaclk:direct=0\na=group:DUP ra0 ra1\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.99.240.1/16\na=source-filter: incl IN IP4 239.99.240.1 192.168.99.240\na=rtpmap:98 L24/48000/2\na=mid:ra0\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.98.240.1/16\na=source-filter: incl IN IP4 239.98.240.1 192.168.98.240\na=rtpmap:98 L24/48000/2\na=mid:ra1\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\n
```

I also tried creating a trigger to automatically fill the variable from the SDP field of the device's Ember+ Provider, but that only resulted in a single line variable with neither line breaks nor `\n`. If I paste the original SDP into the Power Core using Ember+ Viewer, the Power Core accepts it.

Best regards,

Christoffer

**Comment by phillipivan** [comment]
G'Day Christoffer,

Thanks for coming back with more info. I don't think any of the devices I have access to handle multi-line strings so it's not something I have been able to test directly.

The module itself doesn't do anything special with the strings supplied to it, other than parsing variables, so I wonder if this is a characteristic of the ember-plus library being used.

Can I send you a module with some extra logging to test with?

**Comment by cbjornram** [comment]
Absolutely, if I can help then I would like to.

As additional information, I ran a Wireshark after sending my previous message that I think shows where the issue might be. When patching an SDP with Ember+ Viewer, there are carrage returns (hex 0d 0a) as linebreaks throughout the SDP.

![Image](https://github.com/user-attachments/assets/738bfab1-17b1-4c41-a48b-adce802f896a)

However, the SDP patching message from Companion/Generic Ember Plus module contains the `\n` characters. so it seems as if the `\n` characters are not parsed to real linebreaks before being sent.

![Image](https://github.com/user-attachments/assets/67bb718a-ba1d-4e5e-a9f7-d66791c0534c)

**Comment by phillipivan** [comment]
@cbjornram here is a build with some logging (at console level) inserted after parsing of string variables, it should confirm what you have observed in wireshark:

```
case EmberModel.ParameterType.String:
	value = await self.parseVariablesInString(action.options['value']?.toString() ?? '')
	console.log(`Set Value String:\n${value}`)
	break
```

[companion-module-generic-emberplus_v262.zip](https://github.com/user-attachments/files/18564724/companion-module-generic-emberplus_v262.zip)

**Comment by phillipivan** [comment]
[companion-module-generic-emberplus_v263.zip](https://github.com/user-attachments/files/18569186/companion-module-generic-emberplus_v263.zip)
Give this a try a well if you can
@cbjornram is there an powercore ember+ emulator or similar that I could use to test with?

**Comment by cbjornram** [comment]
I configured a button with "PowerCore: Set Value String" as the only action on it as the following image. As can be seen in the tiny button preview, the SDP gets rendered correctly with linebreaks in the Companion UI.

![Image](https://github.com/user-attachments/assets/6062f6dd-3ed5-45ba-a1f6-9c46cf3f73bc)

The $(custom:sdptest) variable was set to:

```
v=0\no=- 09420142300001 0 IN IP4 192.168.99.240\ns=core01:rav01\nt=0 0\na=clock-domain:PTPv2 0\na=ts-refclk:ptp=IEEE1588-2008:00-0B-72-FF-FE-09-88-F6:0\na=mediaclk:direct=0\na=group:DUP ra0 ra1\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.99.240.1/16\na=source-filter: incl IN IP4 239.99.240.1 192.168.99.240\na=rtpmap:98 L24/48000/2\na=mid:ra0\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.98.240.1/16\na=source-filter: incl IN IP4 239.98.240.1 192.168.98.240\na=rtpmap:98 L24/48000/2\na=mid:ra1\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\n
```

Pressing it on version 262 gave the following output:

```
system: Starting log. Only lines generated since opening the page are shown here
debug: Got node on 1.6.3.2.1.6
Set Value String:
v=0\no=- 09420142300001 0 IN IP4 192.168.99.240\ns=core01:rav01\nt=0 0\na=clock-domain:PTPv2 0\na=ts-refclk:ptp=IEEE1588-2008:00-0B-72-FF-FE-09-88-F6:0\na=mediaclk:direct=0\na=group:DUP ra0 ra1\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.99.240.1/16\na=source-filter: incl IN IP4 239.99.240.1 192.168.99.240\na=rtpmap:98 L24/48000/2\na=mid:ra0\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.98.240.1/16\na=source-filter: incl IN IP4 239.98.240.1 192.168.98.240\na=rtpmap:98 L24/48000/2\na=mid:ra1\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\n
```

Pressing it on version 263 gave the same output:

```
system: Starting log. Only lines generated since opening the page are shown here
debug: Got node on 1.6.3.2.1.6
Set Value String:
v=0\no=- 09420142300001 0 IN IP4 192.168.99.240\ns=core01:rav01\nt=0 0\na=clock-domain:PTPv2 0\na=ts-refclk:ptp=IEEE1588-2008:00-0B-72-FF-FE-09-88-F6:0\na=mediaclk:direct=0\na=group:DUP ra0 ra1\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.99.240.1/16\na=source-filter: incl IN IP4 239.99.240.1 192.168.99.240\na=rtpmap:98 L24/48000/2\na=mid:ra0\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\nm=audio 5004 RTP/AVP 98\nc=IN IP4 239.98.240.1/16\na=source-filter: incl IN IP4 239.98.240.1 192.168.98.240\na=rtpmap:98 L24/48000/2\na=mid:ra1\na=framecount:48\na=recvonly\na=ptime:1\na=sync-time:0\na=mediaclk:direct=0\n
```

I did verify that the correct version was picked up by Companion before creating the logs. I also tested that that same button still can clear SDPs by sending a blank string, which it could with the following log output:

```
system: Starting log. Only lines generated since opening the page are shown here
debug: Got node on 1.6.3.2.1.6
Set Value String:

```


Unfortenately, there are no Power Core emulators that I'm aware of. If it's of any help, the Ember+ Viewer does show and send linebreaks of strings properly, so if you have any provider available (for example the provider of Companion itself), you can write to button labels and see the result. Here I have used shift+return to set a linebreak of a Companion button while inside the "Change..." dialogue:

![Image](https://github.com/user-attachments/assets/f7ae078b-53a4-49e9-83a4-1b92c9dbe050)

**Comment by phillipivan** [comment]
Good idea to test with Companion's own ember plus provider. I'll have another look next week.

**Comment by phillipivan** [comment]
I have observed the same issue with a syslog client module I wrote a few days ago, so I suspect there is a common root cause.

**Comment by phillipivan** [comment]
HI @cbjornram @fgoupil-cbc please test this build. It seems to be working for me - controlling companion button text from companion. Using ember+ viewer to confirm the field value.

The fix will only work for `\r` `\n` and `\t`. It is something of a hack, and I have reservations about pushing it to a proper release as the problem lies elsewhere, but for the time being I think it will get you going.

[companion-module-generic-emberplus_v263.zip](https://github.com/user-attachments/files/18918567/companion-module-generic-emberplus_v263.zip)

**Comment by cbjornram** [comment]
Hi Phillip, thank for the status updates and the new build. I will give this a try as soon as I can (which might be a few days/a week away). I'll keep you posted.

**Comment by fgoupil-cbc** [comment]
Thanks Phillip and sorry for now being able to participate earlier. 

I ran a quick test and it stlll doesn't seem to work with my Powercore. Here are some details :

- Companion v2.5.2
- I took the zip package you send and replaced the bundled Generic Ember+ module.
- Created a custom variable containing the SDP file and added the line breaks (\n) for each line + an extra one at the end (required in Ember+ viewer otherwise the SDP is not accepted).
- Created a button with emberplus: Set Value String
- I entered my path (1.6.3.2.7.6) and the custom value.
- Powercore still doesn't register the SDP.

Should I be testing with another version of Companion? I validated the path by emptying the SDP in the Powercore and it works.

**Comment by phillipivan** [comment]
Hi @fgoupil-cbc can you confirm the module version is reporting as 2.6.3 when you open the module help?

**Comment by phillipivan** [comment]
Testing against companion with module 2.6.3 (and Companion 3.5.2)
This action:

![Image](https://github.com/user-attachments/assets/d0938367-9b6f-4f1c-8047-dd1abfba480f)
Results in this value according to ember plus viewer:

![Image](https://github.com/user-attachments/assets/c4e10e76-07e0-4fd4-8fcb-abfaa6fcd26b)

If I revert to 2.6.2 and perform the same action, it appears as:

![Image](https://github.com/user-attachments/assets/7dfb079b-3180-4a8b-93d9-f8fa597a0a0d)

Can you confirm you get the same results.

If this is still not working with the Powercore then I am not sure where to be looking.

**Comment by fgoupil-cbc** [comment]
Hello Phillip,

I did label test and I do get line breaks in the button with version 2.6.3 and not when I backtrack to 2.6.2 so that part is working. However, the Powercore still doesn't like the way the SDP is formatted (tried \n and \r). I tried reaching out to Lawo but I was not able to get a useful response from them... I'll try to dig this further at some point but it seems like the Powercore is very picky about how the SDP is formatted. One thing I could try is catch an SDP advertisement being sent from the Powercore and see how it's formatted - I'll report back when I get a chance to do it.

Thanks,

François

**Comment by phillipivan** [comment]
Out of interest, @fgoupil-cbc are you able to successfully set other parameters on the power core? Like a mute or fader level or channel name?

**Comment by fgoupil-cbc** [comment]
Yes @phillipivan, I'm able to control boolean driven controls successfully from Companion. I tried \r\n and no luck either for the SDP. It's too bad the Powercore is so picky on SDP formatting since for simple setups, Companion would be a great way to control streams going into a Powercore.

**Comment by phillipivan** [comment]
Yes I understand.

I am hoping @cbjornram chimes back in with some insight as it looks like he works at Lawo.

There is one other thing I can try - updating the core ember-plus library the module uses. I have been avoiding this since the last time I tried it completely broke control with the devices I use; I have been meaning to follow that up but haven't got to it yet (in no small part since the older version seems OK). If that does help I guess I should direct some more attention at it.

**Comment by phillipivan** [comment]
@fgoupil-cbc here is a build using the most recent [emberplus library](https://github.com/nrkno/sofie-emberplus-connection/releases/tag/v0.3.0). As mentioned this version causes issues with the devices I work with, but maybe it will help here. Let me know if it helps.

[companion-module-generic-emberplus_v264.zip](https://github.com/user-attachments/files/18999281/companion-module-generic-emberplus_v264.zip)

**Comment by phillipivan** [comment]
I have one last idea to try - which admittedly is a bit of a Hail Mary.

If you subscribe to the ember path, ie with a feedback; you can then engage the action recorder. If you then change the SDP in ember plus viewer or another control system it will create a set value string action with the data received. There is an outside chance this might pick up some difference that is being missed otherwise.

I have no idea if this is more likely to work with 2.6.2 or the custom builds I have shared - but it should only take 15m to find out.

**Comment by fgoupil-cbc** [comment]
@phillipivan I did try the 2.6.4 version and it causes Ember+ issues with the Powercore so I had to revert to 2.6.3. I'll give a spin with the action recorder tomorrow.

Thanks!

**Comment by phillipivan** [comment]
2.7.0 is now published and available here. Please let me know if this helps.

https://github.com/bitfocus/companion-module-generic-emberplus/releases/tag/v2.7.0

**Comment by phillipivan** [comment]
I have a WIP PR that updates the core ember plus library. If I post a build can you test with that?

It would also be useful to try setting the action value with the learn button.
